### ISSUE TITLE
One microservice or two, that is the question

### ISSUE THREAD

**Issue by elkdanger** [opencastsoftware/leave-tracker#1]
Let's kick off development of this thing with one of the questions I've been rattling around in my head for a few days. It's something we touched on in the initial discussions but it's the question of whether we go with a frontend and a backend, or just one simple app that serves the frontend as well as doing all the business logic.

**tl,dr; I'm leaning towards just having one app for MVP.**

I'll outline some possibilites here.

**Option 1** there was some talk about an Angular frontend and a Scala API to provide the data.

Pros:

* Separates things nicely
* More sustainable long-term
* Good practise
* Might get a nice UX

Cons:

* Might be over-engineering for what we need at this stage
* Only one or two people have Angular experience, although I don't think the actual framework was fully decided
* Makes things like Auth a lot harder
* More moving parts to go wrong
* Higher hosting costs, etc (might not really be an issue)

**Option 2:** One app that serves a server-side front-end and all the business logic

Pros:

* Simpler to implement and run
* Will get MVP faster
* Don't have to worry about how we're going to authenticate API calls into the backend
* Could still implement a front-end framework

Cons:

* It's not the future
* Risk of developing a monolith
* Everyone has to be on board with the tech
* Implementing a frontend framework, while still possible, would be messy (and it kind of rules out Angular, unless we can think of a clean way to integrate it)

I think right now that Option 2 is our best way to get started but absolutely believe that we can (and should) transition to Option 1 later once we get the app off the ground and people start using it, if there is a requirement for it. I just think that otherwise we might be over-engineering the solution when it might not really be needed at this stage.

Thoughts? I figured once we decide on which route to go, we could start to think about the tech and get some initial stories drawn up.

**Comment by Froom2** [comment]
I agree, I had a similar discussion with Jack when I was starting with my pet project and although I resisted furiously, he made good points - as an mvp it's fine.  This is why I ended up using Firebase for that pet project.
I'm not saying Firebase is the solution for us, because I don't know how easy it would be to do the office 365 thing, but it's the same reasoning... what we're building isn't huge, and isn't really going to get overwhelming levels of traffic.
If things get super busy we may never have the time to build anything more than the basic frontend.  I'd say there is a real case for something even simpler like Firebase/Angular, but as you say, there's less knowledge on that in the group so it makes sense to go down option 2 route i think. Let's get something of value working and running super fast :D

**Comment by msultan2** [comment]
I agree with you that option2 is simpler but it will take longer and some code reworks in the transition phase from option2 to option1. So I think making it in the right way with cleaner code better than trying to separate it afterwards with workarounds and other similar problems.
So I vote for option1 without thinking of the advanced functionalities like Auth as I don't think it should be covered in the MVP for the moment.

**Comment by JGRobinson** [comment]
I think that you are essentially talking about creating a controller service which should then call other smaller grained services in order to provide the behaviour. I think a happy medium would be to create the controller service, and then Inject the other behaviours in - that would then be easier to separate later, and you wouldn't get side tracked with things like security claim propagation and schema separation.

**Comment by elkdanger** [comment]
@msultan2 The problem with not considering Auth up front is that we still need to have some way to identify who is interacting with the system so we can create data against the right user.

@nonimay and I have been batting some ideas about using Office365 OAuth or OpenID, which would tie in to our current organisation, and would be fairly simple. There's an issue coming to discuss that idea in more detail.

**Comment by JGRobinson** [comment]
+1 for O365 OAuth ....

**Comment by elkdanger** [comment]
@JGRobinson For Option 2 I'm simply talking about a single app that runs in one process that contains all the business logic + serves the front end, rather than simply providing an API for another frontend to interact with. We can absolutely develop the code in a modular way which would make it easier to separate later if required, and that would just be down to implementation detail (and something I hope we would do anyway as a matter of course!).

As mentioned there are trade-offs to both options, but which one is going to get us something up and running, in a sensible fashion, that we can iterate on later?

**Comment by JGRobinson** [comment]
Ok - then you have to accept that it would be largely throw away which doesn’t feel right for a problem where we largely know the correct answer (the MVP being Jim’s Spreadsheet)

**Comment by Froom2** [comment]
I'm okay with throwing away code :)  The next iteration might be better.  This is the whole point of agile, right?  Build an actual MVP get something out fast that works and gives value, throw it away later if necessary. We _think_ we know the right answer (a digitisation of Jim's spreadsheet) but that's kind of putting the cart before the horse a little bit. Build something small, put it out there, see what the users think, then build from that based on user need.

**Comment by JGRobinson** [comment]
As long as everyone is onboard with that ethos

**Comment by elkdanger** [comment]
Absolutely on board. 'Knowing the answer' from Jim's spreadsheet urges me towards option 2 even more, because one thing it does tell us is that we're building an MVP for 10's of occasional users in one organisation; it doesn't need to scale to 1000's of concurrent users and be multi-tenant.

I can imagine a lot of the frontend will be throwaway if we choose to break the front-end out later on, but if we build the business logic in such a way that we can just put a thin API on top of it later, then we could continue using a lot of that stuff.

**Comment by JGRobinson** [comment]
If thats the case then, would it be better to focus more on the front end mock data into it etc for a few iterations to inform the backend more, rather than worry about services at all at this point?

Maybe even use the lean startup approach  (https://www.amazon.co.uk/Lean-Startup-Innovation-Successful-Businesses-ebook/dp/B005PR422K/ref=tmm_kin_swatch_0?_encoding=UTF8&qid=&sr=) which basically are quick iteration, backed up by data and experiments?

**Comment by elkdanger** [comment]
@JGRobinson Absolutely, for some of it at least. A lot of the data will be user-generated anyway, but there will come a time where we will need to populate it with historical data, or data that Jim already knows about for the forthcoming year, which we can definitely mock until such time that we need to import it.

Not familiar with the lean startup approach though - can you give us a quick summary?

**Comment by JGRobinson** [comment]
I'll try :-)

So the general approach is that you try and workout what you are trying to prove - these are called experiments.

An example may be that there is a requirement that holiday is approved and the "customer" is informed of the outcome within 24 hours (as a random requirement)

You would then decide what data you would need to collect, to access whether or not it was a success.

You would then come up with some hypothesis on how you would fulfil the requirement, and carry out the bare minimum to test that hypothesis (may not even be code) ... so for instance there are many ways of messaging people - mail, slack, skype ... probably others so the MVP in this instance may be to send the messages by hand to different people and compare the results.

You would then iterate on the results with any additional learning leaving the actual cutting of the code til the last responsible minute ... thus reducing waste.

This is just a basic example, and things are never as clear cut in the real world, but its an interesting technique which takes the emotion out of decisions as any decisions made are always backed up by data.

It's a really fascinating book to read - i tend to read it once a year, and see something different every time.

**Comment by elkdanger** [comment]
@JGRobinson Definitely warrants some further reading, thanks for that.

In the interests of getting this moving, I propose that we go ahead and create an app in this very repository which will serve as our MVP. I think the next stage is to draw up a set of issues that act as epics that we can then break down into a set of stories that we can work on (I can do that based on some initial requirements we went through with Jon Everett after the company meeting), which will all be up for discussion and review.

I have a few rough UI sketches that I'll share once we've fleshed out the stories a little bit.

Finally, I think at the last count we'd more or less floated the idea of building this as a Play Scala app. There's additional motivation for doing so as it will serve as a learning and training tool. Unless anyone has a strong opinion as to why we shouldn't be doing that, we'll run with that for now. The frontend will be nothing special for now but as per this discussion, may warrant revisiting after MVP has been delivered.
